utils/core/src/serde/byte_reader.rs

```diff
@@ -0,0 +1,166 @@
+use super::{Deserializable, DeserializationError};
+
+// BYTE READER TRAIT
+// ================================================================================================
+
+/// Defines how primitive values are to be read from `Self`.
+pub trait ByteReader {
+    // REQUIRED METHODS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a single byte read from `self`.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] error the reader is at EOF.
+    fn read_u8(&mut self) -> Result<u8, DeserializationError>;
+
+    /// Returns the next byte to be read from `self` without advancing the reader to the next byte.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] error the reader is at EOF.
+    fn peek_u8(&self) -> Result<u8, DeserializationError>;
```

Comment:
optional nit: peek could instead return an `Option` as the user might just want to check if there is another byte to be read. The signature would look like:

```rust
fn peek_u8(&self) -> Result<Option<u8>, DeserializationError>;
```

Also, another method that can be helpful is
```rust
fn peek_into(&self, buf: &mut [u8]) -> Result<usize, DeserializationError>;
```

Without this method, we cannot peek more than one byte, and this might be a too restrictive

Reply:
For the first one, we already have `has_more_bytes()` method. The second one would be cool - but I'm also leaving it for the future.